### ISSUE TITLE
RediSearch prevents repl-diskless-load

### ISSUE THREAD

**Issue by tw-bert** [RediSearch/RediSearch#1955]
When setting `repl-diskless-load` to `on-empty-db` or `swapdb` , redis log mentions, after a `slaveof` to a valid replication source:

```log
19603:S 18 Apr 2021 00:01:52.057 # Skipping diskless-load because there are modules that don't handle read errors.
```

**iff** redisearch is in the modules. We commented out all other modules.
After the log message (repeated numerous times), redis-server switches back to using disk (dump.rdb, temp-dump etc).

We tried the following RediSearch versions on the replica target (we didn't change the replica source, we couldn't see if that was a necessity but could be wrong in assuming it's not...):

- `1.4.10` (yeah, it's an oldie, but used by our largest lan-to-wan stream atm)
- `1.6.15` , the latest 1.6 as of now (remark: we think the upgrade path from 1.4 to 1.6 can be done without rebuilding the FT indexes, but if not so, rebuilding and migrating to FT 2.0.6+ is a bit of work but not undoable)
- `2.0.6`, the latest 2.x as of now (which we absolutely love)

For redis-server, we're at `6.2.1`.
 
So it seems this is currently not supported by RediSearch.
Is it somewhere on the roadmap?

It would help us out a lot. Currently, the disk I/O performance on our cloud provider is appaling. A full sync of a 16GB redis instance takes about 40 minutes. And that's about the same as just doing a `bgsave` on the cloud VM itself. The limiting factor is the disk performance. Network is fine.

There are other options for us, like switching providers, or going SSD on cloud (which we are considering as well).

But since redis is in-memory, and we have plenty of canaries and hot failovers / redundancy to be confident enough to use this feature, it would be great if we could use diskless load for full syncs.

Also let me know if I can assist in any way.

Looking forward to your reply, cheers, TW

**Comment by MeirShpilraien** [comment]
Hey @tw-bert we already have an issue for that:
https://github.com/RediSearch/RediSearch/issues/1682
Maybe @emmanuelkeller or @rafie can give more information about timelines.

**Comment by tw-bert** [comment]
@MeirShpilraien  @emmanuelkeller @rafie : Any news? Anything I can do to assist?

**Comment by rafie** [comment]
@tw-bert This will become available in the next patch version, v2.0.9, a few weeks from now, and obviously in master once the fix is ready.

**Comment by tw-bert** [comment]
Hi @rafie , I tested with v2.0.9, but I still get the same no-go.

replication target log:

```
23829:S 15 Jun 2021 12:53:36.895 # Skipping diskless-load because there are modules that don't handle read errors.
```

**Comment by tw-bert** [comment]
@MeirShpilraien @emmanuelkeller @rafie : Anything I can do to help out? 
If it's in the master, I could test, but the issue (for us) is only relevant for production cloud VM's (so no unstables).
But please tell me if a test on master will help you in any way.
Cheers + TIA, TW

**Comment by rafie** [comment]
@tw-bert PR #2058 will be merged into master soon and will be available in v2.0.10, if everything goes according to plan.

**Comment by tw-bert** [comment]
@rafie Great, thank you for the heads-up

**Comment by tw-bert** [comment]
@rafie I think I see the PR didn't make the cut for v2.0.10?

**Comment by MeirShpilraien** [comment]
@tw-bert this feature ended up more complex than we originally thought (especially the testing part), we are working on it, stay tune.

**Comment by rafie** [comment]
@tw-bert Our implementation exposed a problem in Redis (https://github.com/redis/redis/pull/9199 - already fixed and merged). This was vital for a correct implementation. Thus, we'll need to be checking in RediSearch for a minimum Redis version(s) that contains the fix, so we're waiting for the next Redis patch for such Redis version to become available.

**Comment by tw-bert** [comment]
@rafie Fantastic, good sleuthing ;) 

We'll await the next Redis release, and after that, the new RediSearch release. Thank you for the update, all clear.

**Comment by tw-bert** [comment]
@rafie I see a lot of effort went into this, much obliged. Now it is merged to master, is there an ETA for a release?

**Comment by tw-bert** [comment]
@MeirShpilraien see previous post

**Comment by MeirShpilraien** [comment]
@emmanuelkeller any estimation when we plan to get it in?

**Comment by MeirShpilraien** [comment]
@tw-bert seems like currently its only target to 2.2. We will see how much work and how risky it is to cherypick. Regardless, notice that we will have to do more work to support it in future Redis versions after this will get it
https://github.com/redis/redis/pull/9323

**Comment by tw-bert** [comment]
@MeirShpilraien Thank you for the update(s), all clear. For our purposes, cherry picking into v2.0 will definately help (if safe/feasible).

**Comment by tw-bert** [comment]
@rafie @MeirShpilraien @emmanuelkeller Bump.

I have to ask, our company will have to decide on what to do.

From my 1st post in this thread:

> There are other options for us, like switching providers, or going SSD on cloud (which we are considering as well). 

Not trying to push, if this functionality is out of reach we simply have to choose a sub-optimal but 'good enough' approach, and that's fine.

**Comment by MeirShpilraien** [comment]
@tw-bert 2.2 should be released next week (the RC1 should even be released today)

**Comment by tw-bert** [comment]
@MeirShpilraien Just an update, thanks for the 2.2 release. We'll plan some tests, I'll keep you posted.

**Comment by tw-bert** [comment]
@oshadmi @rafie @MeirShpilraien @emmanuelkeller
We migrated our stage environments from FT 1.4.10 to 2.2.6 two days ago, drop and redump of a dozen indexes inside a 20GB instance, which went smooth.
We haven't encountered a single issue during regression tests. 
Diskless replication works as intended. 
Thank you all very much for this functionality! 
Closing this issue.
